### ISSUE TITLE
StableDiffusionXLImg2ImgPipeline: denoising_start doesn't work correctly with 2nd order schedulers

### ISSUE THREAD

**Issue by nhnt11** [huggingface/diffusers#5447]
### Describe the bug

We were setting up SDXL with the refiner in an ensemble of experts configuration, where we split the generation between the base model and the refiner.

When using `denoising_start` for the refiner step, we noticed that the result images were either completely black, or not fully denoised for order=2 schedulers (heun, dpm2, dpm2 a).

Some investigation revealed that `get_timesteps` is not returning the right slice of the scheduler's timesteps which seems to be causing the issue. When replacing the logic with the exact same logic as used by the `denoising_end` calculation in the prompt-to-image pipeline code, but just inverting the slicing, we got the expected results.

Here is the working `get_timesteps` function we used:
```python3
    def get_timesteps(self, num_inference_steps, strength, device, denoising_start=None):
        # get the original timestep using init_timestep
        if denoising_start is None:
            init_timestep = min(int(num_inference_steps * strength), num_inference_steps)
            t_start = max(num_inference_steps - init_timestep, 0)
        else:
            t_start = 0

        timesteps = self.scheduler.timesteps[t_start * self.scheduler.order :]

        # Strength is irrelevant if we directly request a timestep to start at;
        # that is, strength is determined by the denoising_start instead.
        if denoising_start is not None:
            discrete_timestep_cutoff = int(
                round(
                    self.scheduler.config.num_train_timesteps
                    - (denoising_start * self.scheduler.config.num_train_timesteps)
                )
            )
            num_timesteps = len(list(filter(lambda ts: ts >= discrete_timestep_cutoff, timesteps)))
            timesteps = timesteps[len(timesteps) - num_timesteps:]
            return timesteps, len(timesteps)

        return timesteps, num_inference_steps - t_start
```

### Reproduction

This script results in a black image with stock diffusers. With the change we suggest, we get the expected result.

```python3
import torch
from diffusers import StableDiffusionXLPipeline, StableDiffusionXLImg2ImgPipeline
from typing import cast
from diffusers import HeunDiscreteScheduler

sdxl_model = cast(StableDiffusionXLPipeline, StableDiffusionXLPipeline.from_pretrained(
    'stabilityai/stable-diffusion-xl-base-1.0',
    torch_dtype=torch.float16,
    use_safetensors=True,
    variant="fp16",
    revision="76d28af79639c28a79fa5c6c6468febd3490a37e",
)).to('cuda')
refiner_model = cast(StableDiffusionXLImg2ImgPipeline, StableDiffusionXLImg2ImgPipeline.from_pretrained(
    'stabilityai/stable-diffusion-xl-refiner-1.0',
    torch_dtype=torch.float16,
    use_safetensors=True,
    variant="fp16",
    revision="5d4cfe854c9a9a87939ff3653551c2b3c99a4356",
)).to('cuda')

common_config = {'beta_start': 0.00085, 'beta_end': 0.012, 'beta_schedule': 'scaled_linear'}
scheduler = HeunDiscreteScheduler(**common_config)
sdxl_model.scheduler = scheduler
refiner_model.scheduler = scheduler

sdxl_model.watermark = None
generator = torch.Generator(device='cuda')
generator.manual_seed(12345)

hnf = 0.5
params = {
    'prompt': ['evening sunset scenery blue sky nature, glass bottle with a galaxy in it'],
    'negative_prompt': ['text, watermark'],
    "negative_prompt": [''],
    "num_inference_steps": 50,
    "height": 1024,
    "width": 1024,
    "guidance_scale": 7,
}

sdxl_res = sdxl_model(**params, denoising_end=hnf, generator=generator, output_type='latent')
sdxl_latents = sdxl_res.images

refiner_res = refiner_model(
    output_type="pil",
    image=sdxl_latents,
    prompt=params['prompt'],
    num_inference_steps=params['num_inference_steps'],
    negative_prompt=params['negative_prompt'],
    generator=generator,
    denoising_start=hnf,
)
refiner_imgs = refiner_res.images

display(refiner_imgs[0])
```

### Logs

With stock diffusers, we see this in the output:
```
[...]/site-packages/diffusers/image_processor.py:88: RuntimeWarning: invalid value encountered in cast
  images = (images * 255).round().astype("uint8")
```

### System Info

- `diffusers` version: 0.21.4
- Platform: Linux-5.4.0-163-generic-x86_64-with-glibc2.31
- Python version: 3.11.5
- PyTorch version (GPU?): 2.1.0+cu121 (True)
- Huggingface_hub version: 0.17.1
- Transformers version: 4.34.0
- Accelerate version: 0.22.0
- xFormers version: not installed
- Using GPU in script?: <fill in>
- Using distributed or parallel set-up in script?: <fill in>

### Who can help?

@yiyixuxu @sayakpaul @DN6 @patrickvonplaten

**Comment by sayakpaul** [comment]
Cc: @yiyixuxu @patrickvonplaten could you look into it?

**Comment by nhnt11** [comment]
OK our solution does not work with all values of hnf and denoising_start, so I dug a bit further. It seems that the black images are caused by a single timestep being omitted at the start of the timesteps list. I suspect it's caused by the rounding in this line:
```python3
            discrete_timestep_cutoff = int(
                round(
                    self.scheduler.config.num_train_timesteps
                    - (denoising_start * self.scheduler.config.num_train_timesteps)
                )
            )
```
